spotbugs/src/main/java/edu/umd/cs/findbugs/detect/DateFormatStringChecker.java

```diff
@@ -0,0 +1,183 @@
+
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.StringAnnotation;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+public class DateFormatStringChecker extends OpcodeStackDetector {
+    private final BugReporter bugReporter;
+
+    /**
+     * Contains special flags that can trigger check (keywords); property of the rule
+     * (isRequired - whether flag should contain or be absent); flags to be checked
+     * and a bypassing flag (whether a rule can be skipped if such flag is included)
+     */
+    private static final class Rule {
+        boolean isRequired;
+        String ignore;
+        List<String> flags;
+        List<String> keywords;
+
+        Rule(List<String> flags, String ignore, boolean isRequired, List<String> keywords) {
+            this.flags = flags;
+            this.ignore = ignore;
+            this.isRequired = isRequired;
+            this.keywords = keywords;
+        }
+
+        boolean containsAny(String target, List<String> list) {
+            if (list == null) {
+                return false;
+            }
+            for (String item : list) {
+                if (target.contains(item)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        boolean verify(String dateFormat) {
+            if (!containsAny(dateFormat, this.keywords)
+                    || (this.ignore != null && dateFormat.contains(this.ignore))) {
+                return false;
+            }
+
+            boolean cond = containsAny(dateFormat, this.flags);
+            if (this.isRequired) {
+                return !cond;
+            }
+            return cond;
+        }
+    }
+
+    public DateFormatStringChecker(BugReporter bugReporter) {
+        this.bugReporter = bugReporter;
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (!(seen == Const.INVOKESPECIAL || seen == Const.INVOKEVIRTUAL || seen == Const.INVOKESTATIC
+                || seen == Const.INVOKEINTERFACE) || stack.getStackDepth() == 0) {
+            return;
+        }
+
+        int idx = -1;
+        String cl = getClassConstantOperand();
+        String nm = getNameConstantOperand();
+        String si = getSigConstantOperand();
+        if ("java/text/SimpleDateFormat".equals(cl)
+                && contains(nm, Const.CONSTRUCTOR_NAME, "applyPattern", "applyLocalizedPattern")) {
+            switch (si) {
+            case "(Ljava/lang/String;)V":
+                idx = 0;
+                break;
+            case "(Ljava/lang/String;Ljava/util/Locale;)V":
+            case "(Ljava/lang/String;Ljava/text/DateFormatSymbols;)V":
+                idx = 1;
+                break;
+            }
+        } else if ("java/time/format/DateTimeFormatter".equals(cl) && "ofPattern".equals(nm)) {
+            switch (si) {
```

Comment:
Can you please provide a few lines of comments to the switches, why that is the value of `idx`?

Reply:
fixed